aspnet/mvc/models/file-uploads.rst

```diff
@@ -0,0 +1,197 @@
+
+.. literalinclude:: file-uploads/sample/FileUploadSample/Controllers/ProfileController.cs
+  :language: c#
+  :lines: 46-64
+  :dedent: 8
+  :emphasize-lines: 2,10,13
+
+Uploading Large Files with Streaming
+------------------------------------
+
+If the size or frequency of file uploads is causing performance or resource problems for the app, consider streaming the file upload rather than buffering it in its entirety (as the model binding approach shown above does). Using ``IFormFile`` and model binding is a much simpler solution; streaming requires a number of steps to implement properly.
+
+The following example demonstrates how to using JavaScript/Angular to stream to a controller action. The file's antiforgery token will be generated using a custom filter attribute, and then passed in HTTP headers instead of in the request body. Because the action method will be processing the uploaded data directly, model binding will be disabled by another filter. Within the action, the form's contents are read using a MultipartReader, which reads each individual MultipartSection, processing the file or storing the contents as appropriate. Once all sections have been read, the action performs its own model binding.
+
+The initial action loads the form and saves an antiforgery token in a cookie (via the ``GenerateAntiforgeryTokenCookieForAjax`` attribute):
+
+.. literalinclude:: file-uploads/sample/FileUploadSample/Controllers/StreamingController.cs
+  :language: c#
+  :lines: 36-41
+  :emphasize-lines: 2
+
+The attribute uses ASP.NET Core's built-in :doc:`Antiforgery </security/anti-request-forgery>` support to set a cookie with a request token:
+
+.. literalinclude:: file-uploads/sample/FileUploadSample/Filters/GenerateAntiforgeryTokenCookieForAjaxAttribute.cs
+  :language: c#
+  :emphasize-lines: 16-19
+
+Angular automatically passes an antiforgery token in a request header named ``X-XSRF-TOKEN``; the ASP.NET Core MVC app is configured to refer to this header in its configuration in ``Startup.cs``:
+
+.. literalinclude:: file-uploads/sample/FileUploadSample/Startup.cs
+  :language: c#
+  :lines: 10-16
+  :emphasize-lines: 3-4
+
+The ``DisableFormValueModelBinding`` attribute is used to disable model binding for the ``Upload`` action method:
+
+.. literalinclude:: file-uploads/sample/FileUploadSample/Filters/DisableFormValueModelBindingAttribute.cs
+  :language: c#
+  :emphasize-lines: 18,26
+
+Since model binding is disabled, the ``Upload`` action method doesn't accept any parameters. It works directly with the ``Request`` property of ``ControllerBase``. A ``MultipartReader`` is used to read each section, and either save the file (with a GUID filename) or accumulate the key/value data in a ``KeyValueAccumulator``. Once all sections have been read, the contents of the ``KeyValueAccumulator`` are used to bind the form data to a model type.
```

Comment:
Why don't we preserve the file extension when saving the file?

"and **either** save or ..." -- it actually does both, not either.  Suggest something like this:  "A ``MultipartReader`` is used to read each section. The file is saved with a GUID filename and the key/value data is saved in a ``KeyValueAccumulator``."